chunk_reader.go

```diff
@@ -0,0 +1,127 @@
+package smtp
+
+import (
+	"errors"
+	"io"
+	"io/ioutil"
+)
+
+// ErrDataReset is returned by Reader pased to Data function if client does not
+// send another BDAT command and instead closes connection or issues RSET command.
+var ErrDataReset = errors.New("smtp: message transmission aborted")
+
+// chunkReader implements io.Reader by consuming a sequence of size-limited
+// "chunks" from underlying Reader allowing them to be interleaved with other
+// protocol data.
+//
+// It is caller responsibility to not use Reader while chunk is being processed.
+// This can be enforced by blocking on chunkEnd channel that is used to signal the
+// end of another chunk being reached.
+type chunkReader struct {
+	remainingBytes int
+	conn           io.Reader
```

Comment:
What do you thing about renaming `conn` to `r`, `reader`, or `input`?

Reply:
Done